stdlib/buffer.gr

```diff
@@ -0,0 +1,747 @@
+/**
+ * @module Buffer: Utilities for working with buffers.
+ *
+ * Buffers are an automatically growing data structure useful for storing and operating on an unknown number of bytes. When working with buffers, all set or append operations will be mutating.
+ * @example import Buffer from "buffer"
+ *
+ * @since v0.4.0
+ */
+
+import Memory from "runtime/unsafe/memory"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import Conv from "runtime/unsafe/conv"
+import Exception from "runtime/exception"
+import Int32 from "int32"
+import Bytes from "bytes"
+import String from "string"
+import { coerceNumberToWasmI32 } from "runtime/numbers"
+
+record Buffer {
+  mut len: Number,
+  initialSize: Number,
+  mut data: Bytes,
+}
+
+let _SIZE_OFFSET = 4n;
+
+let _VALUE_OFFSET = 8n;
+
+let _8BIT_LEN = 1
+
+let _16BIT_LEN = 2
+
+let _32BIT_LEN = 4
+
+let _64BIT_LEN = 8
+
+/* Gets the size of a Bytes via its ptr */
+@disableGC
+let getSize = (ptr) => WasmI32.load(ptr, _SIZE_OFFSET)
+
+/* Doubles the size of buffer's underlying byte sequence, if the given size is larger than the size of a buffer's underlying byte sequence */
+let autogrow = (len, buf) => {
+  while (buf.len + len > Bytes.length(buf.data)) {
+    let mut n = Bytes.length(buf.data)
+    if (n == 0) n = 4 // Make sure bytes of 0 length grow too
+    buf.data = Bytes.resize(0, n, buf.data)
+  }
+}
+
+/* Gets the pointer for a char's bytes following the char's tag */
+@disableGC
+let rec getCharAsWasmI32 = (char) => {
+  let c = WasmI32.fromGrain(char)
+  WasmI32.load(c, 4n)
+}
+
+/* Gets the UTF-8 byte length of a char */
+@disableGC
+let rec getCharByteLength = (byte) => {
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (==) = WasmI32.eq
+  if ((byte & 0x80n) == 0x00n) {
+    1n
+  } else if ((byte & 0xF0n) == 0xF0n) {
+    4n
+  } else if ((byte & 0xE0n) == 0xE0n) {
+    3n
+  } else {
+    2n
+  }
+}
+
+/* Memcopies bytes from a source byte sequence to a destination byte sequence via pointers */
+@disableGC
+let rec appendBytes = (srcOff, dstOff, len, src, dst) => {
+  let (+) = WasmI32.add
+  Memory.copy(dst + _VALUE_OFFSET + dstOff, src + _VALUE_OFFSET + srcOff, len)
+}
+
+/*
+ * If given index is negative, returns a positive index.
+ * Throws if a given index + len is not within the bounds of a buffer's current length
+ */
+let getSafeIndex = (index, len, buf) => {
+  let i = index % buf.len
+  if (i >= buf.len || i + len > buf.len) throw Exception.IndexOutOfBounds
+  i
+}
```

Comment:
Ping @ospencer and @peblair - are we good/happy to apply the logic of "a negative index is the index from the end" to Buffers?

Reply:
@jozanza depending on the answer here, we may need some negative index tests.

Reply:
Yes, that's correct.

Reply:
We decided that negative indexing doesn't make sense for Buffer because the data isn't constant size, like an array or list.